WooCommerce/Classes/ViewRelated/Products/Variations/ProductVariationsViewModel.swift

```diff
@@ -26,27 +26,49 @@ final class ProductVariationsViewModel {
     }
 
     /// Generates all missing variations for a product. Up to 100 variations.
+    /// Parameters:
+    /// - `Product`: Product on which we will be creating the variations
+    /// - `onStateChanged`: Closure invoked every time there is a significant state change in the generation process.
     ///
-    func generateAllVariations(for product: Product, onCompletion: @escaping (Result<Void, GenerationError>) -> Void) {
+    func generateAllVariations(for product: Product, onStateChanged: @escaping (GenerationState) -> Void) {
 
+        // Fetch Previous variations
+        onStateChanged(.fetching)
         fetchAllVariations(of: product) { [weak self] result in
-            guard let self else { return }
             switch result {
             case .success(let existingVariations):
 
+                // Generate variations locally
                 let variationsToGenerate = ProductVariationGenerator.generateVariations(for: product, excluding: existingVariations)
 
                 // Guard for 100 variation limit
                 guard variationsToGenerate.count <= 100 else {
-                    return onCompletion(.failure(.tooManyVariations(variationCount: variationsToGenerate.count)))
+                    return onStateChanged(.error(.tooManyVariations(variationCount: variationsToGenerate.count)))
                 }
 
+                // Guard for no variations to generate
                 guard variationsToGenerate.count > 0 else {
-                    // TODO: Inform user that no variation will be created
-                    return onCompletion(.success(()))
+                    return onStateChanged(.finished(false))
                 }
 
-                self.createVariationsRemotely(for: product, variations: variationsToGenerate, onCompletion: onCompletion)
+                // Confirm generation with merchant
+                onStateChanged(.confirmation(variationsToGenerate.count, { confirmed in
```

Comment:
This one is hard to read and understand, but I don't have easy improvement suggestions 🤔

Reply:
Yeah, not a fan either. will keep it in the back of my head for improvements!